servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererObserverTest.java

```diff
@@ -107,9 +113,38 @@ void srvQueryTriggersNewDiscoveryObserver() throws Exception {
     private void testNewDiscoveryObserver(BiFunction<DnsClient, String, Publisher<?>> publisherFactory,
                                           String expectedName) throws Exception {
         BlockingQueue<String> newDiscoveryCalls = new LinkedBlockingDeque<>();
-        DnsClient client = dnsClient(name -> {
-            newDiscoveryCalls.add(name);
-            return NoopDnsDiscoveryObserver.INSTANCE;
+        final AtomicBoolean discoveryCanceledCalled = new AtomicBoolean();
+        final AtomicBoolean discoveryFailedCalled = new AtomicBoolean();
+        Queue<Throwable> errors = new LinkedBlockingQueue<>();
+        DnsClient client = dnsClient(new DnsServiceDiscovererObserver() {
+            @Override
+            public DnsDiscoveryObserver onNewDiscovery(final String name) {
+                errors.add(new AssertionError("This method must not be called anymore when overridden"));
+                return ignored -> NoopDnsResolutionObserver.INSTANCE;
+            }
+
+            @Override
+            public DnsDiscoveryObserver onNewDiscovery(final String serviceDiscovererId, final String name) {
+                newDiscoveryCalls.add(name);
+                assertThat("Unexpected serviceDiscovererId", serviceDiscovererId, equalTo(DISCOVERER_ID));
```

Comment:
This assertion will not fail the test because we always try-catch observer callbacks internally. Consider using `errors` queue. You can create your own `AssertionError` or try-catch `assertThat`.